### ISSUE TITLE
Regex.Match fails for the .* pattern when startat is non-zero

### ISSUE THREAD

**Issue by slozier** [dotnet/runtime#42390]
### Description

Regex.Match fails for `.*` pattern when `startat` is non-zero. I would expect the following to succeed but it does not.

```C#
Debug.Assert(new Regex(".*").Match("abc", startat: 1).Success);
```

### Configuration

.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.1.20452.10
 Commit:    473d1b592e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-rc.1.20452.10\

Host (useful for support):
  Version: 5.0.0-rc.1.20451.14
  Commit:  38017c3935

### Regression?

This works in .NET Core 3.1 and .NET Framework.

**Comment by stephentoub** [comment]
This is due to https://github.com/dotnet/runtime/pull/1706.  That PR adds an optimization that automatically adds a beginning-of-line anchor to a pattern that begins with `.*`.  This lets the implementation skip backtracking by jumping to the start of the next line when it fails to match.  This, however, didn't factor in the possibility that matching might be starting in the middle of a line, in which case when it goes to match, the BOL node itself may fail to match.

The simple fix is to just delete the optimization, i.e. remove this code:
https://github.com/dotnet/runtime/blob/a747feefaa3b0c5c7e33ee42ba94e00927e54160/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs#L353-L404

That removes a valuable optimization, but I'm not sure if there's a good way to save it.  One thought is to add a new kind of anchor, one that behaves like BOL for the purposes of jumping ahead in FindFirstChar but that doesn't actually require any matching in Go.  I expect that would work, but it's probably too late to do so for .NET 5.

@pgovind, @danmosemsft, thoughts?

**Comment by jeffhandley** [comment]
Fixed in #42408 (master) and #42409 (release/5.0)
